program/src/sigverify.rs

```diff
@@ -0,0 +1,185 @@
+#[repr(C)]
+pub(crate) struct Ed25519SignatureOffsets {
+    pub(crate) signature_offset: u16, // offset to ed25519 signature of 64 bytes
+    pub(crate) signature_instruction_index: u16, // instruction index to find signature
+    pub(crate) public_key_offset: u16, // offset to public key of 32 bytes
+    pub(crate) public_key_instruction_index: u16, // instruction index to find public key
+    pub(crate) message_data_offset: u16, // offset to start of message data
+    pub(crate) message_data_size: u16, // size of message data
+    pub(crate) message_instruction_index: u16, // index of instruction data to get message data
+}
+
+impl Ed25519SignatureOffsets {
+    pub(crate) fn to_instruction(offsets: &[Self]) -> Instruction {
+        let mut instruction_data = Vec::with_capacity(
+            SIGNATURE_OFFSETS_START
+                .saturating_add(SIGNATURE_OFFSETS_SERIALIZED_SIZE.saturating_mul(offsets.len())),
+        );
+
+        let num_signatures: u8 = offsets.len() as u8;
+
+        // add padding byte so that offset structure is aligned
+        instruction_data.extend_from_slice(bytemuck::bytes_of(&[num_signatures, 0]));
+
+        for offsets in offsets {
+            instruction_data.extend_from_slice(bytemuck::bytes_of(offsets));
+        }
+
+        Instruction {
+            program_id: ed25519_program::id(),
+            accounts: vec![],
+            data: instruction_data,
+        }
+    }
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub(crate) struct SignatureVerification<'a> {
+    pub(crate) pubkey: &'a Pubkey,
+    pub(crate) message: &'a [u8],
+    pub(crate) signature: &'a [u8; SIGNATURE_BYTES],
+}
+
+impl<'a> SignatureVerification<'a> {
+    fn new(
+        pubkey: &'a [u8],
+        message: &'a [u8],
+        signature: &'a [u8],
+    ) -> Result<SignatureVerification<'a>, SlashingError> {
+        let pubkey: &'a Pubkey = bytemuck::try_from_bytes(pubkey).map_err(|_| {
+            msg!("Failed to deserialize pubkey");
+            SlashingError::InvalidSignatureVerification
+        })?;
+
+        let signature: &'a [u8; SIGNATURE_BYTES] = signature.try_into().map_err(|_| {
+            msg!("Failed to deserialize signature");
+            SlashingError::InvalidSignatureVerification
+        })?;
+
+        Ok(Self {
+            pubkey,
+            message,
+            signature,
+        })
+    }
+
+    fn get_data_slice<'b>(
+        data: &'a [u8],
+        _instructions_sysvar: &'a AccountInfo<'b>,
+        instruction_index: u16,
+        current_index: u16,
+        offset_start: u16,
+        size: usize,
+    ) -> Result<&'a [u8], SlashingError> {
+        if instruction_index != current_index {
+            // For duplicate block slashing the message is small enough to fit inside the
+            // slashing instruction but this might not be the case for future slashing cases
+            // TODO: re-implement load_instruction_at_checked(instruction_index as usize,
```

Comment:
I can take care of this in a future PR when we get to vote transactions, it seems pretty straight forward to reimplement the offset lookups.